### PR TITLE
feat(design): rework modal to be dynamically rendered

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently, using a modal requires a convoluted portal system.

Fixes: N/A

## What is the new behavior?
This uses dynamic rendering to the portal the modal without having to setup up additional components.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

Any existing application that uses the old modal will no longer have backdrops. This is a hard break, and any apps would need updating.

## Other information

@lderrickable @xelaint This will have subsequent PRs to restore pre-existing functionality, like closing the modal. Additionally, it will follow up with better action hygiene. I didn't want this PR to become unmanageable so, I did the smallest amount of code I could to upgrade.

### CODE REVIEW

File: libs/design/src/molecules/backdrop/backdrop.module.ts [hunk 1]
```diff
@@ -12,6 +12,9 @@ import { DaffBackdropComponent } from './backdrop/backdrop.component';
   ],
   exports: [
     DaffBackdropComponent
+  ],
+  entryComponents: [
+    DaffBackdropComponent
```
Comment (lderrickable) [inline libs/design/src/molecules/backdrop/backdrop.module.ts]: should be a different PR

**Comment by lderrickable** [comment]
The build fails in a big way, so I'm going to wait to review it further till it's a little more tidy

File: apps/demo/src/app/cart/components/add-to-cart-notification/effects/add-to-cart-notification.effects.ts [hunk 2]
```diff
@@ -1,19 +1,36 @@
...
+  private notification: DaffModal;
+
+  constructor(private store$: Store<any>, private daffModalService: DaffModalService) {}
     
-  @Effect()
-  addToCart$ = this.actions$.pipe(
-    ofType(DaffCartActionTypes.AddToCartAction),
-    map(() => new OpenAddToCartNotification())
+  @Effect({dispatch: false})
+  showModalWhenNotificationIsOpen$ = this.store$.pipe(
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply apps/demo/src/app/cart/components/add-to-cart-notification/effects/add-to-cart-notification.effects.ts]: Then for the "close button clicked" action you could do:
```

	
  @Effect({dispatch: false})
  closeModal$ = this.actions$.pipe(
		ofType(AddToCartNotificationActionTypes.CloseButtonClicked),
    map((action) => {
      this.daffModalService.close(this.notification)
    })
  )
```

File: apps/design-land/src/app/modal/modal.module.ts [hunk 3]
```diff
@@ -1,24 +1,31 @@
...
-    DaffFeatureModule
-
+    DaffFeatureModule,
+    DaffCardModule
+  ],
+  entryComponents: [
+    DesignLandModalContentComponent
   ]
 })
 export class ModalModule {
```
Comment (xelaint) [inline apps/design-land/src/app/modal/modal.module.ts]: Prefix with `DesignLand` ?

File: apps/design-land/src/app/modal/modal.component.ts [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
 
-  constructor() { }
-
-  ngOnInit() {
+  showModal() {
+    this.modal = this.modalService.open(DesignLandModalContentComponent);
   }
-  private showModal : boolean;
+
   hideModal() {
```
Comment (lderrickable) [inline apps/design-land/src/app/modal/modal.component.ts]: This function doesn't do anything

File: apps/design-land/src/app/modal/modal-content/modal-content.component.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+})
+export class DesignLandModalContentComponent {
+  private modal: any;
+  constructor(private modalService: DaffModalService) {}
+
+  showModal() {
+    this.modal = this.modalService.open(DesignLandModalContentComponent);
+  }
+
+  hideModal() {
```
Comment (lderrickable) [inline apps/design-land/src/app/modal/modal-content/modal-content.component.ts]: maybe add an "x" to the modal content to demonstrate this? Or just remove this function

File: apps/demo/src/app/cart/components/add-to-cart-notification/effects/add-to-cart-notification.effects.ts [hunk 6]
```diff
@@ -1,19 +1,46 @@
...
+import { AddToCartNotificationComponent } from '../components/add-to-cart-notification/add-to-cart-notification.component';
 
 @Injectable()
 export class AddToCartNotificationEffects {
+  private notification: DaffModal;
 
-  constructor(private actions$: Actions) {}
+  constructor(
+    private actions$: Actions,
+    private daffModalService: DaffModalService,
```
Comment (lderrickable) [inline apps/demo/src/app/cart/components/add-to-cart-notification/effects/add-to-cart-notification.effects.ts]: The DaffModalModule needs to be imported somewhere. The service isn't being provided to demo right now.

File: apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.html [hunk 7]
```diff
@@ -1,21 +1,20 @@
...
+      <fa-icon [icon]="faTimes" aria-hidden="true"></fa-icon>
+    </button>
+  </div>
+  <demo-product-added *ngIf="!(loading$ | async)" class="demo-add-to-cart-notification__product-added"
+    [product]="product$ | async" [qty]="productQty$ | async"></demo-product-added>
+  <div *ngIf="!(loading$ | async)" daff-button-set class="demo-add-to-cart-notification__button-set">
+    <button type="button" daff-stroked-button demoViewCart>
+      View Cart ({{(cartItemCount$ | async)}})
+    </button>
+    <button type="button" daff-button color="secondary" demoProceedToCheckout>Checkout</button>
```
Comment (lderrickable) [inline apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.html]: clicking either of these buttons should also close the modal.

File: apps/design-land/src/app/modal/modal.module.ts (hunk 3 above)
Comment (damienwebdev) [reply apps/design-land/src/app/modal/modal.module.ts]: fixed in cf659493bb6334ae3b97c3125b3d6f4f6ab27cef

File: libs/design/src/molecules/modal/modal/modal.component.ts [hunk 8]
```diff
@@ -1,84 +1,91 @@
...
+	 */
+	@Output() closedAnimationCompleted: EventEmitter<any> = new EventEmitter<
+		any
+	>();
+
+	/**
+	 * Event fired when the backdrop is clicked
+	 * This is often used to close the modal
+	 */
+	@Output() hide: EventEmitter<void> = new EventEmitter<void>();
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/design/src/molecules/modal/modal/modal.component.ts]: I listen to it in the ModalService, I can remove it as an Output.

File: apps/design-land/src/app/modal/modal.component.ts (hunk 4 above)
Comment (damienwebdev) [reply apps/design-land/src/app/modal/modal.component.ts]: Fixed in b75505d2cbbe48d71a604ee3a1eb1a5c597ebc24

File: apps/design-land/src/app/modal/modal-content/modal-content.component.ts (hunk 5 above)
Comment (damienwebdev) [reply apps/design-land/src/app/modal/modal-content/modal-content.component.ts]: Fixed in b75505d2cbbe48d71a604ee3a1eb1a5c597ebc24

File: apps/demo/src/app/cart/components/add-to-cart-notification/effects/add-to-cart-notification.effects.ts (hunk 6 above)
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/add-to-cart-notification/effects/add-to-cart-notification.effects.ts]: Fixed in b75505d2cbbe48d71a604ee3a1eb1a5c597ebc24

File: apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.html (hunk 7 above)
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.html]: I plan on addressing this in the next PR. See my note in the PR description:

> @lderrickable @xelaint This will have subsequent PRs to restore pre-existing functionality, like closing the modal. Additionally, it will follow up with better action hygiene. I didn't want this PR to become unmanageable so, I did the smallest amount of code I could to upgrade.

File: libs/design/src/molecules/modal/modal/modal.component.ts [hunk 9]
```diff
@@ -1,84 +1,91 @@
...
 
 import { daffFadeAnimations } from '../animations/modal-animation';
 import { getAnimationState } from '../animations/modal-animation-state';
+import { CdkPortalOutlet, ComponentPortal } from '@angular/cdk/portal';
+
+import { AnimationEvent } from '@angular/animations';
+import { map } from 'rxjs/operators';
+
+export type DaffModalVerticalPosition = 'top' | 'center' | 'bottom';
+export type DaffModalHorizontalPosition = 'left' | 'center' | 'right';
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/design/src/molecules/modal/modal/modal.component.ts]: Fixed in2da4da15bd4a60a9aabe96c0b4499f95a421739f

File: libs/design/src/molecules/modal/modal/modal.component.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/design/src/molecules/modal/modal/modal.component.ts]: Fixed in 3903ba6e7889e852569d71037df49385fea443c9

**Comment by lderrickable** [comment]
lgtm

File: libs/design/src/molecules/modal/animations/modal-animation.ts [hunk 10]
```diff
@@ -1,23 +1,18 @@
...
-    })),
-    state('void',style({
-      opacity: 0
-    })),
-    transition('void <=> show',
-        animate('350ms cubic-bezier(0.25, 0.8, 0.25, 1)'))
-  ])
+	fade: trigger('fade', [
+		state('open', style({ opacity: 1, })),
+		state('closed', style({ opacity: 0, transform: 'translateY(-20px)' })),
```
Comment (xelaint) [inline libs/design/src/molecules/modal/animations/modal-animation.ts]: Remove transform style

File: libs/design/src/molecules/modal/animations/modal-animation.ts [hunk 11]
```diff
@@ -1,23 +1,18 @@
...
+		transition('open <=> closed', animate('100ms cubic-bezier(0.25, 0.8, 0.25, 1)')),
```
Comment (xelaint) [inline libs/design/src/molecules/modal/animations/modal-animation.ts]: Modal timing differs from backdrop. I think they should be the same. It looks awkward to have backdrop fade out later than the modal.

File: libs/design/src/molecules/modal/modal.md [hunk 12]
```diff
@@ -1,39 +1,44 @@
...
+  entryComponents: [
+    MyModalContentComponent
+  ]
+})
+export class ModalModule { 
+}
 ```
 
-### verticalPosition
+Once your module is setup, you can take advantage of the `DaffModalService` that is provided by the `DaffModalModule` to open and close your instance of the `DaffModalComponent`.
```
Comment (xelaint) [inline libs/design/src/molecules/modal/modal.md]: Once `the`?

File: apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.scss [hunk 13]
```diff
@@ -24,6 +24,7 @@
 
 	&__close {
 		position: absolute;
+		top: 20px;
```
[the remark this answers is not in the document]
Comment (damienwebdev) [reply apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.scss]: Out of scope.

[the remark this answers is not in the document]
Comment (xelaint) [reply apps/demo/src/app/cart/components/add-to-cart-notification/components/add-to-cart-notification/add-to-cart-notification.component.scss]: This file should be removed from the PR then

File: libs/design/src/molecules/modal/modal.md [hunk 14]
```diff
@@ -24,7 +27,7 @@ export class ModalModule {
 }
 ```
 
-Once your module is setup, you can take advantage of the `DaffModalService` that is provided by the `DaffModalModule` to open and close your instance of the `DaffModalComponent`. 
+Once you have configured your `@NgModule` you can take advantage of the `DaffModalService` that is provided by the `DaffModalModule` to `open` and `close` the instance of the `DaffModalComponent`.
```
Comment (xelaint) [inline libs/design/src/molecules/modal/modal.md]: Once you have configured **the** `@NgModule`**,** you can take advantage of the ......

**Comment by xelaint** [comment]
Modal doesn't close if I click `view cart` or `checkout` ?

File: libs/design/src/molecules/modal/modal.md (hunk 14 above)
Comment (damienwebdev) [reply libs/design/src/molecules/modal/modal.md]: I feel "your" is more appropriate given that the user is configuring their own particular module.

**Comment by damienwebdev** [comment]
> Modal doesn't close if I click `view cart` or `checkout` ?

> @lderrickable @xelaint This will have subsequent PRs to restore pre-existing functionality, like closing the modal. Additionally, it will follow up with better action hygiene. I didn't want this PR to become unmanageable so, I did the smallest amount of code I could to upgrade.

File: libs/design/src/molecules/modal/modal.md (hunk 14 above)
Comment (xelaint) [reply libs/design/src/molecules/modal/modal.md]: Once your `@NgModule` has been configured, ....
